### PR TITLE
fix: route path-prefixed commands to shell, preserve backslashes in history

### PR DESCRIPTION
**Pull request by lacymorrow**
## Paperclip Issue
LAC-449

## Summary
- Commands starting with path prefixes (`/`, `./`, `../`, `~/`) were incorrectly routed to the AI agent when `command -v` couldn't verify the binary
- Paths with backslash-escaped spaces (e.g., `/Applications/Google\ Chrome.app/...`) were losing their backslashes in ZSH history

## Root Cause

**Bug 1 — Mis-routing:** `lacy_shell_classify_input` relied entirely on `command -v` to detect executables. When `command -v` failed (binary not installed, path with spaces edge cases), the function fell through to the "multiple words → agent" branch. For a command like `/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --remote-debugging-port=9222`, the arguments after the path are non-empty, so it classified as `agent`.

**Bug 2 — History backslash loss:** ZSH's `print` command (even with `-s`) processes `\X` escape sequences. An unrecognized `\ ` (backslash-space) is silently converted to just ` `. This caused history entries for agent-routed commands to drop backslashes.

## Fix

**`lib/core/detection.sh`:** Added early path-prefix check in `lacy_shell_classify_input` (auto mode only). If the input starts with `/`, `./`, `../`, or `~/`, it immediately routes to shell — no backslash-space extraction or `command -v` needed. This handles executables whose path includes spaces regardless of whether they're installed.

**`lib/zsh/execute.zsh`:** When adding agent-routed commands to ZSH history via `print -s`, double all backslashes first (`\ → \\`). ZSH's `print` then converts `\\` → `\`, preserving the original escaped path in history.

## Acceptance Criteria
- [ ] `/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --remote-debugging-port=9222` routes to shell (green indicator), not agent
- [ ] After the command runs, pressing Up retrieves it with backslashes intact
- [ ] `./script.sh --flag`, `../other/tool`, and `~/bin/tool` all route to shell
- [ ] Normal commands like `ls -la`, `git status`, natural language like `what files are here` still route correctly

## Testing Notes
Source `lacy.plugin.zsh` in a test shell and try the Chrome command. The indicator should turn green (shell) as you type the path. The `lacy_shell_test_detection` function can be used to verify the classification output.

### CODE REVIEW

File: lib/core/detection.sh [hunk 1]
```diff
@@ -122,6 +122,15 @@ lacy_shell_classify_input() {
...
     fi
 
+    # Auto mode: executable path prefixes (/..., ./..., ../..., ~/...) always
+    # route to shell. Handles paths with backslash-escaped spaces (e.g.,
+    # /Applications/Google\ Chrome.app/...) without relying on the
+    # backslash-space substitution below, which behaves differently in ZSH vs Bash.
+    if [[ "$input" == /* || "$input" == ./* || "$input" == ../* || "$input" == '~/'* ]]; then
+        echo "shell"
+        return
+    fi
```
[the remark this answers is not in the document]
Comment (lacymorrow) [reply lib/core/detection.sh]: The broad path-prefix block this comment references was removed in the latest commit (c729991) after review feedback that it was too aggressive —  typed as an agent reply would incorrectly route to shell. The root cause was fixed instead (ZSH `$'\x01'` expansion bug in substitutions). The `~` pattern is moot.

File: lib/zsh/execute.zsh [hunk 2]
```diff
@@ -91,8 +91,12 @@ lacy_shell_smart_accept_line() {
...
             fi
 
-            # Add to history before clearing buffer
-            print -s -- "$input"
+            # Add to history before clearing buffer.
+            # Double backslashes before print -s: ZSH's print processes \X escape
+            # sequences even with -s, so "Google\ Chrome" becomes "Google Chrome".
+            # Doubling (\ → \\) makes print convert \\ → \, preserving the original.
+            local _hist_input="${input//\\/\\\\}"
+            print -s -- "$_hist_input"
```
[the remark this answers is not in the document]
Comment (lacymorrow) [reply lib/zsh/execute.zsh:100]: Tested both `print -s` and `print -S` — both strip backslashes from `\ ` sequences (verified via `xxd`). ZSH's `print` processes `\X` escapes regardless of the uppercase/lowercase `-s/-S` flag; the docs confirm `-s` and `-S` ignore all other flags including `-r`. The backslash-doubling workaround is necessary.

The consistency note for line 30 is valid — fixed in the latest commit (d407ba6): slash commands now also use the backslash-doubling approach for consistency.

**Comment by lacymorrow** [comment]
## Code Review — PR #27 (LAC-449)

### Overview

Two targeted fixes across `lib/core/detection.sh` and `lib/zsh/execute.zsh`. Root cause is correctly identified and the fix is minimal and precise.

---

### `lib/core/detection.sh` — `_lacy_bsp` placeholder fix

**Verdict: Correct.** The root cause is accurately diagnosed — ZSH treats `$'\x01'` inside `${var//pattern/replacement}` as the literal 6-char string, not the ANSI-C SOH byte. Assigning to a variable first forces evaluation at assignment time. This is a well-known ZSH edge case.

- `_lacy_bsp` is declared `local`, so no namespace pollution
- The fix works identically in Bash (ANSI-C quoting at assignment time is standard in both)
- After the fix, `command -v` correctly receives the unescaped path (e.g., `/Applications/Google Chrome`) for lookup
- Removal of the old `if [[ "$input" == /* ]]` early-exit is clean — it was a workaround; the proper `_lacy_bsp` fix makes it redundant

**One note:** The PR description says the fix was an "early path-prefix check" (`if [[ "$input" == /* ]]`). The actual diff uses the `_lacy_bsp` variable approach instead — which is better. **The PR description is stale** and should be updated before merge.

For non-existent paths with arguments (`/nonexistent/bin foo bar`): `command -v` fails and the multi-word fallthrough routes to agent. Single-word non-existent paths still go to shell via the typo path. Both are per design.

---

### `lib/zsh/execute.zsh` — backslash-doubling for `print -s`

**Verdict: Correct.** ZSH's `print -s` processes `\X` escape sequences; `\ ` silently becomes ` `. Doubling (`\ → \\`) before storage makes `print` output `\`, round-tripping correctly. Applied consistently to:

- Agent query path (line ~99) — primary fix
- Slash-command path (line ~30, commit `d407ba6`) — consistency fix

Edge case check: inputs with literal `\\` double-escape correctly and round-trip.

---

### Potential risks

- **No new tests**: `lacy_shell_test_detection` already includes backslash-escaped path cases for the detection logic. History round-trip is harder to automate in shell — acceptable gap.
- **`print -r` alternative**: Would suppress all escape processing without doubling. Not used here; the doubling approach is equivalent and consistent with the rest of the file.

---

### Summary

| Area | Finding |
|------|---------|
| Root cause diagnosis | Correct |
| Fix scope | Minimal — only two affected paths |
| Bash compatibility | Maintained |
| Inline comments | Clear, explain non-obvious ZSH behavior |
| PR description | **Stale — update before merge** |

**Result: Self-review complete. Recommend merge after updating the PR description.**
